### PR TITLE
Azure kms handling for noobaa

### PR DESCRIPTION
**Pull request by v-harihar**
This patch provides the support for azure keyvault.

We are using `libopenstorage/secrets` as the wrapper package to integrate with different kms and package provides the abstraction over several kms.  It also provides the integraton support for azure and helps communication with azure key vault.

We are required to provide the definition for house keeping calls registered calls with libopenstorage/secrets.

`libopenstorage/secrets` does the creation of client handle based on the details provided in config data. Noobaa CR will have these details under `connectionDetails` and these details are popupated on Noobaa CR by Noobaa ocs by reading the configmap provided by ODF. The certificate details present in the secret are copied into a temp file and used to establish the connection with azure key vault as of now.

`libopenstorage/secrets `supports only secret as string as of now. Our requirement is to authenticate using certificate. To facililate this, we have PR to accomodate certificate related handling and the PR can be found here: https://github.com/libopenstorage/secrets/pull/83/files

Below are the connection details that are going to be populated on Noobaa CR by `noobaa_system_reconciler` at ocs side and this is the ocs code where connectiondetails on Noobaa CR are built: https://github.com/red-hat-storage/ocs-operator/blob/2d082fc4c1ac4cec961406053cece448f4b07684/controllers/storagecluster/noobaa_system_reconciler.go#L249

ex: configmap data:
AZURE_CERT_SECRET_NAME = Secret name from which we extract the certificate data and send it to libopenstorage/secrets to get the client handle for service principle to access the root master key
```
data:
  AZURE_CERT_SECRET_NAME: azure-ocs-ffwc9o1j
  AZURE_CLIENT_ID: az-client-id1
  AZURE_TENANT_ID: az-tenant-id1
  AZURE_VAULT_URL: az-valut-url1
  KMS_PROVIDER: azure-kv
  KMS_SERVICE_NAME: kms-conn-azure1
```

### CODE REVIEW

File: pkg/util/kms/kms_azure.go
Comment (jackyalbo) [inline pkg/util/kms/kms_azure.go:44]: Give some info about what is needed by the secrets module. What is this cert file and why did we create it also about the other config variables

**Review by jackyalbo — COMMENTED** [review]
I think we need to add test for azure KMS, also I think we need some more info about what we expect the info sent by the NooBaa CR should look like, I mean what config info we need.

File: pkg/util/kms/kms_ibm_kp_storage.go
Comment (v-harihar) [reply pkg/util/kms/kms_ibm_kp_storage.go:98]: That was unnoticed. I will correct it

File: pkg/util/kms/kms_azure.go
Comment (v-harihar) [reply pkg/util/kms/kms_azure.go:37]: Actually `*AzureVault` is just used to bind it to method and not using it inside the function/method

Comment (jackyalbo) [reply pkg/util/kms/kms_azure.go:37]: I see, I was confused with the key, value (k,v) in line 40

File: pkg/util/kms/kms_azure.go
Comment (jackyalbo) [reply pkg/util/kms/kms_azure.go:44]: but let's remove all the unneeded as well, like in Get/Set/DeleteContext

File: pkg/util/kms/kms_azure.go
Comment (jackyalbo) [reply pkg/util/kms/kms_azure.go:37]: but let's remove all the unneeded as well, like in Get/Set/DeleteContext

File: pkg/util/kms/kms_azure.go
Comment (v-harihar) [reply pkg/util/kms/kms_azure.go:44]: We get below data as part of `config map[string]string` parameter. This details comes from connectiondetails section of noobaa CR (https://github.com/noobaa/noobaa-operator/blob/906a82032bfc8927145186984034c41577467c9d/pkg/util/kms/kms.go#L99)

ex: configmap data:
```
data:
  AZURE_CERT_SECRET_NAME: azure-ocs-ffwc9o1j
  AZURE_CLIENT_ID: az-client-id1
  AZURE_TENANT_ID: az-tenant-id1
  AZURE_VAULT_URL: az-valut-url1
  KMS_PROVIDER: azure-kv
  KMS_SERVICE_NAME: kms-conn-azure1
```

File: pkg/util/kms/kms_azure.go
Comment (v-harihar) [reply pkg/util/kms/kms_azure.go:37]: These are the once we need to interface them with the driver type. Here: https://github.com/noobaa/noobaa-operator/blob/906a82032bfc8927145186984034c41577467c9d/pkg/util/kms/kms.go#L44C1-L53C2

File: pkg/util/kms/kms_azure.go
Comment (jackyalbo) [inline pkg/util/kms/kms_azure.go:49]: from where do we know that we expect a file path and not just the string itself?

File: pkg/util/kms/kms_azure.go
Comment (jackyalbo) [inline pkg/util/kms/kms_azure.go:48]: Does it really has to do with TLS? otherwise, I think we would expect a both key and cert.

File: pkg/util/kms/kms_azure.go
Comment (jackyalbo) [reply pkg/util/kms/kms_azure.go:49]: That is what I found in the secrets: https://github.com/libopenstorage/secrets/blob/5f4b25ceede180e8d67577288bba4223af658eb1/azure/azure_kv.go#L64
Am I missing something?

File: pkg/util/kms/kms_azure.go
Comment (jackyalbo) [inline pkg/util/kms/kms_azure.go:90]: Not sure if it has to do with TLS, I would expect both key and cert in that case. maybe you just need to update the comment...

**Review by jackyalbo — COMMENTED** [review]
I have some comment, the main issue I have is with AZURE_CERT_SECRET_NAME
which I don't understand if it needs to be in a file and where this requirement is coming for, what I did find in the secrets was AZURE_CLIENT_SECRET which I'm not sure if that't what you meant, and it looks like a regular string.

**Comment by v-harihar** [comment]
> I have some comment, the main issue I have is with AZURE_CERT_SECRET_NAME which I don't understand if it needs to be in a file and where this requirement is coming for, what I did find in the secrets was AZURE_CLIENT_SECRET which I'm not sure if that't what you meant, and it looks like a regular string.

There are 2 things. 
1. libopenstorage/secrets currently support "secret" as string. 
2. The requirement is to go for "secret" as certificate. 

Currently libopenstorage/secrets dont support secret as certificate. We have (Member from Rook team, Santosh, working for azure kms on rook side and owning this [azure kms story](https://issues.redhat.com/browse/RHSTOR-4799))  pushed a patch to  libopenstorage/sectes support secret certificate and that can be found here: https://github.com/libopenstorage/secrets/pull/83/files#diff-91153976f4824b52d20fd598108e77d2a5d605f6b67dedec4f0ba53e70ca5e49R82-R93

With the certificate handling in place, we need to create a temp file and put the temp path in the map. This is used at libopenstrorage/secrets to create a service principle. The code for the same is here at: https://github.com/libopenstorage/secrets/pull/83/files#diff-feaeae3297b3a2250b55e5307f4f8ef82bf0c6032af4ffe0736c6814ec7855f5R33-R37

The path we store in the map is accessed at the libopenstorage/secrets at here: https://github.com/libopenstorage/secrets/pull/83/files#diff-91153976f4824b52d20fd598108e77d2a5d605f6b67dedec4f0ba53e70ca5e49R74

I have imported the hash for the same in this PR which  here: https://github.com/noobaa/noobaa-operator/pull/1311/files#diff-33ef32bf6c23acb95f5902d7097b7a1d5128ca061167ec0716715b0b9eeaa5f6R5-R8

File: pkg/util/kms/kms_azure.go
Comment (v-harihar) [reply pkg/util/kms/kms_azure.go:48]: In azure vault, as per the requirement we dont support secret as string. Only providing support for secret as certificate. Did I answer that ?

File: pkg/util/kms/kms_azure.go
Comment (v-harihar) [reply pkg/util/kms/kms_azure.go:90]: Yes. will update the comment. Thats misleading.

**Comment by jackyalbo** [comment]
> > I have some comment, the main issue I have is with AZURE_CERT_SECRET_NAME which I don't understand if it needs to be in a file and where this requirement is coming for, what I did find in the secrets was AZURE_CLIENT_SECRET which I'm not sure if that't what you meant, and it looks like a regular string.
> 
> There are 2 things.
> 
> 1. libopenstorage/secrets currently support "secret" as string.
> 2. The requirement is to go for "secret" as certificate.
> 
> Currently libopenstorage/secrets dont support secret as certificate. We have (Member from Rook team, Santosh, working for azure kms on rook side and owning this [azure kms story](https://issues.redhat.com/browse/RHSTOR-4799)) pushed a patch to libopenstorage/sectes support secret certificate and that can be found here: https://github.com/libopenstorage/secrets/pull/83/files#diff-91153976f4824b52d20fd598108e77d2a5d605f6b67dedec4f0ba53e70ca5e49R82-R93
> 
> With the certificate handling in place, we need to create a temp file and put the temp path in the map. This is used at libopenstrorage/secrets to create a service principle. The code for the same is here at: https://github.com/libopenstorage/secrets/pull/83/files#diff-feaeae3297b3a2250b55e5307f4f8ef82bf0c6032af4ffe0736c6814ec7855f5R33-R37
> 
> The path we store in the map is accessed at the libopenstorage/secrets at here: https://github.com/libopenstorage/secrets/pull/83/files#diff-91153976f4824b52d20fd598108e77d2a5d605f6b67dedec4f0ba53e70ca5e49R74
> 
> I have imported the hash for the same in this PR which here: https://github.com/noobaa/noobaa-operator/pull/1311/files#diff-33ef32bf6c23acb95f5902d7097b7a1d5128ca061167ec0716715b0b9eeaa5f6R5-R8

I'm a bit worried about us relying on a PR that is not yet merged, and not even properly reviewed...
Also think that maybe we should wait for the tests and testing infra before pushing this.
@liranmauda @nimrod-becker what do you think?

**Comment by v-harihar** [comment]
> > > I have some comment, the main issue I have is with AZURE_CERT_SECRET_NAME which I don't understand if it needs to be in a file and where this requirement is coming for, what I did find in the secrets was AZURE_CLIENT_SECRET which I'm not sure if that't what you meant, and it looks like a regular string.
> > 
> > 
> > There are 2 things.
> > 
> > 1. libopenstorage/secrets currently support "secret" as string.
> > 2. The requirement is to go for "secret" as certificate.
> > 
> > Currently libopenstorage/secrets dont support secret as certificate. We have (Member from Rook team, Santosh, working for azure kms on rook side and owning this [azure kms story](https://issues.redhat.com/browse/RHSTOR-4799)) pushed a patch to libopenstorage/sectes support secret certificate and that can be found here: https://github.com/libopenstorage/secrets/pull/83/files#diff-91153976f4824b52d20fd598108e77d2a5d605f6b67dedec4f0ba53e70ca5e49R82-R93
> > With the certificate handling in place, we need to create a temp file and put the temp path in the map. This is used at libopenstrorage/secrets to create a service principle. The code for the same is here at: https://github.com/libopenstorage/secrets/pull/83/files#diff-feaeae3297b3a2250b55e5307f4f8ef82bf0c6032af4ffe0736c6814ec7855f5R33-R37
> > The path we store in the map is accessed at the libopenstorage/secrets at here: https://github.com/libopenstorage/secrets/pull/83/files#diff-91153976f4824b52d20fd598108e77d2a5d605f6b67dedec4f0ba53e70ca5e49R74
> > I have imported the hash for the same in this PR which here: https://github.com/noobaa/noobaa-operator/pull/1311/files#diff-33ef32bf6c23acb95f5902d7097b7a1d5128ca061167ec0716715b0b9eeaa5f6R5-R8
> 
> I'm a bit worried about us relying on a PR that is not yet merged, and not even properly reviewed... Also think that maybe we should wait for the tests and testing infra before pushing this. @liranmauda @nimrod-becker what do you think?

Quick update to this. Rook is planning to have `libopenstorage/secrets` as a submodule (cloned) inside rook and merge the PR (https://github.com/rook/secrets/pull/1) for azure certificate as of now and use it for the tech preview. Once the PR is merged at `libopenstorage/secrets` then we switch to that package. 

The cloned branch (submodule at rook: https://github.com/rook/secrets) is `libopenstorage/secrets` master + PR (for azure certificate handling: https://github.com/rook/secrets/pull/1)

@liranmauda @jackyalbo @nimrod-becker @dannyzaken

**Comment by sp98** [comment]
> I'm a bit worried about us relying on a PR that is not yet merged, and not even properly reviewed... Also think that maybe we should wait for the tests and testing infra before pushing this. @liranmauda @nimrod-becker what do you think?

Agree that we don't have it reviewed by the maintainers yet.  Pinged them a few weeks ago and pinged on slack, but no response yet. 

We have review comments from the our own team members that have been addressed. 

For rook, we created a fork and plan to use that fork. This is temporary until the PR is reviewed by the library maintainer and merged. 

Since the changes were made for `azure` module, I don't see it affecting the other modules (like hashicorp vault).  But we should ensure that there are no issue is manual testing of this fork. We are ensuring that in Rook.

**Review by jackyalbo — APPROVED** [review]

**Pull request merged**
